Update and improve documentation (#45)
exasol/error-reporting-python: doc/user_guide/user_guide.rst

.. _user_guide:

:octicon:`person` User Guide
============================

> This project contains a Python library for describing Exasol error messages.
> This library lets you define errors with a uniform set of attributes.
> Furthermore, the error message is implemented to be parseable,
> so that you can extract an error catalog from the code.

Reviewer: ```suggestion
This project contains a Python library for describing error messages in Exasol Python projects.
This library lets you define errors with a uniform set of attributes.
Furthermore, the error message is implemented to be parseable,
so that you can extract an error catalog from the code.
```

Reviewer 2: I am not 100% happy with _describing_.
* How about _defining_ 
* or _defining, creating, using, and displaying_, though I'm aware, that the actual _display_ is probably up to the application rather than the ERP project.

Maybe the attribute _unique_ is also important here and should be mentioned?

Error Objects
~~~~~~~~~~~~~

Error objects are built using the function :code:`ExaError`.

> Please keep in mind that error-code should satisfy the error-code format (see `code`_).

Reviewer: ```suggestion
Please keep in mind that an error-code should satisfy the error-code format (see `code`_).
```

        "E-TEST-1", "A trivial error", "No mitigation available", {}
    )

Specify Multiple Mitigations
----------------------------

.. code-block:: python

    from exasol import error

    error2 = error.ExaError(
        "E-TEST-2",
        message="Fire in the server room",
        mitigations=[
            "Use the fire extinguisher",
            "Flood the room with halon gas (Attention: make sure no humans are in the room!)"
        ],
        parameters={}
    )

Error Parameter(s) without description
--------------------------------------

.. code-block:: python

    from exasol import error

    error3 = error.ExaError(
        "E-TEST-2",
        "Not enough space on device {{device}}.",
        "Delete something from {{device}}.",
        {"device": "/dev/sda1"},
    )

Error with detailed Parameter(s)
--------------------------------

.. code-block:: python

    from exasol import error
    from exasol.error import Parameter

    error4 = error.ExaError(
        "E-TEST-2",
        "Not enough space on device {{device}}.",
        "Delete something from {{device}}.",
        {"device": Parameter("/dev/sda1", "name of the device")},
    )

> Check out the `user guide <doc/user_guide/user_guide.md>`_ for more details.

Reviewer 2: A great piece of engineer's prosa - thanks @Nicoretti

Tooling
~~~~~~~

The :code:`exasol-error-reporting` library comes with a command line tool (:code:`ec`) which also can be invoked by using its package/module entry point (:code:`python -m exasol.error._cli`). For detailed information about the usage consider consulting the help :code:`ec --help` or :code:`python -m exasol.error._cli --help`.

Parsing the error definitions in a python file(s)
-------------------------------------------------

.. code-block:: shell

    ec parse some-python-file.py 

.. code-block:: shell

    ec parse < some-python-file.py 

Generating an error-code data file
----------------------------------

> In order to generate a `error-code-report <https://schemas.exasol.com/error_code_report-1.0.0.json>`_ compliant data file, you can use the generate subcommand.

Reviewer 2: ```suggestion
In order to generate an `error-code-report <https://schemas.exasol.com/error_code_report-1.0.0.json>`_ compliant data file, you can use subcommand :code:`generate`:
```
